libr/bin/p/bin_mach0.c

```diff
@@ -387,49 +385,21 @@ static RBinImport *import_from_name(RBin *rbin, const char *orig_name, HtPP *imp
 }
 
 static RList *imports(RBinFile *bf) {
-	RBinObject *obj = bf ? bf->o : NULL;
-	struct MACH0_(obj_t) *bin = bf ? bf->o->bin_obj : NULL;
-	const char *name;
-	RBinImport *ptr = NULL;
-	int i;
-
-	if (!obj || !bin || !obj->bin_obj) {
-		return NULL;
-	}
-	RList *ret = r_list_newf((RListFree)r_bin_import_free);
-	struct import_t *imports = MACH0_(get_imports)(bf->o->bin_obj);
-	if (!ret || !imports) {
-		r_list_free (ret);
-		free (imports);
+	RBinObject *obj = bf? bf->o: NULL;
+	const RPVector *imports = MACH0_(load_imports) (bf, obj->bin_obj);
+	if (!imports) {
 		return NULL;
 	}
-	bin->has_canary = false;
-	bin->has_retguard = -1;
-	bin->has_sanitizers = false;
-	bin->has_blocks_ext = false;
-	for (i = 0; !imports[i].last; i++) {
-		if (!(ptr = import_from_name (bf->rbin, imports[i].name, NULL))) {
-			break;
-		}
-		name = ptr->name;
-		ptr->ordinal = imports[i].ord;
-		if (bin->imports_by_ord && ptr->ordinal < bin->imports_by_ord_size) {
-			bin->imports_by_ord[ptr->ordinal] = ptr;
-		}
-		if (!strcmp (name, "__stack_chk_fail") ) {
-			bin->has_canary = true;
-		}
-		if (!strcmp (name, "__asan_init")
-				|| !strcmp (name, "__tsan_init")) {
-			bin->has_sanitizers = true;
-		}
-		if (!strcmp (name, "_NSConcreteGlobalBlock")) {
-			bin->has_blocks_ext = true;
-		}
-		r_list_append (ret, ptr);
+
+	RList *list = r_list_newf ((RListFree) r_bin_import_free);
+	void **it;
+	r_pvector_foreach (imports, it) {
+		// need to clone here, in bobj.c the list free function is forced to `r_bin_import_free`
+		// otherwise, a list with no free function could be returned here..
```

Comment:
This would also reuse the memory of the vector, which is sequential in memory => better performance

Reply:
Yes but this would break abi, and maybe it will be better to cache the list, but i dont think thats really a bottleneck. Another option is to just use rlist internally and just return the pointer, returned list shouldnt be freed/owned

Reply:
I don't see what would break it? The hack in `bobj.c` needs to be removed, and a free function needs to be added to the returned lists of other file formats (where needed).

Once we have this, we can just use `r_list_new`, with pointers pointing into the vector. Then `r_list_free` would do no freeing except for the list pointer itself, and the vector would manage the other memory.